packages/uma-auth-client/src/hooks/useOAuth.tsx

```diff
@@ -0,0 +1,203 @@
+import { auth } from "@getalby/sdk";
+import { getNwcUmaDomain } from "src/utils/getNwcUmaDomain";
+import { getUmaDomain } from "src/utils/getUmaDomain";
+import { create } from "zustand";
+import { persist } from "zustand/middleware";
+
+interface DiscoveryDocument {
+  authorization_endpoint: string;
+  code_challenge_methods_supported: string[];
+  grant_types_supported: string[];
+  supported_nwc_commands: string[];
+  token_endpoint: string;
+  uma_major_versions: number[];
+  uma_request_endpoint: string;
+}
+
+type UmaAuthToken = {
+  access_token: string;
+  refresh_token: string;
+  expires_at: number;
+  nwc_connection_uri: string;
+  commands: string[];
+  budget: string;
+  nwc_expires_at: number;
+};
+
+interface TokenState {
+  accessToken: string;
+  refreshToken: string;
+  expiresAt: number;
+}
+
+interface OAuthState {
+  codeVerifier?: string;
+  token?: TokenState | undefined;
+  identityNpub?: string;
+  identityRelayUrl?: string;
+  redirectUri?: string;
+  nwcConnectionUri?: string;
+  commands?: string[];
+  budget?: string;
+  nwcExpiresAt?: number;
+  setToken: (token?: TokenState) => void;
+  /** The initial OAuth request that starts the OAuth handshake. */
+  initialOAuthRequest: (uma: string) => Promise<void>;
+  /** OAuth token exchange occurs after the initialOAuthRequest has been made. */
+  oAuthTokenExchange: (uma: string) => Promise<void>;
+}
+
+export const useOAuth = create<OAuthState>()(
+  persist(
+    (set, get) => ({
+      identityNpub:
+        "npub1scmpzl2ehnrtnhu289d9rfrwprau9z6ka0pmuhz6czj2ae5rpuhs2l4j9d", // TODO: remove this
+      identityRelayUrl: "wss://nos.lol", // TODO: remove this
+      redirectUri: "http://localhost:3001", // TODO: remove this
+      setToken: (token) =>
+        set({
+          token,
+        }),
+      initialOAuthRequest: async (uma) => {
+        const state = get();
+        const { codeVerifier, authUrl } = await initialOAuthUrl(state, uma);
+        set({ codeVerifier });
+        window.location.href = authUrl;
+      },
+      oAuthTokenExchange: async (uma) => {
+        const state = get();
+        const result = await oAuthTokenExchange(state, uma);
+        set(result);
+      },
+    }),
+    {
+      name: "uma-connect",
+      partialize: (state) => ({
+        nwcConnectionUri: state.nwcConnectionUri,
+        codeVerifier: state.codeVerifier,
+      }),
+    },
+  ),
+);
+
+const initialOAuthUrl = async (state: OAuthState, uma: string) => {
```

Comment:
updated to getOAuthUrl
it also will get the code verifier here to save in state

Reply:
ehh I still don't think oauthUrl is specific enough because there are multiple "oauth urls". This is specifically the authorization URL as opposed to the token url.